pymc3/distributions/multivariate.py

```diff
@@ -83,55 +47,84 @@ class MvNormal(Continuous):
+        tau, or chol is needed.
 
     """
 
-    def __init__(self, mu, cov=None, tau=None, gpu_compat=False, *args, **kwargs):
+    def __init__(self, mu, cov=None, tau=None, chol=None, *args, **kwargs):
         super(MvNormal, self).__init__(*args, **kwargs)
-        self.mean = self.median = self.mode = self.mu = mu = tt.as_tensor_variable(mu)
-        tau, cov = get_tau_cov(mu, tau=tau, cov=cov)
-        self.tau = tt.as_tensor_variable(tau)
-        self.cov = tt.as_tensor_variable(cov)
-        self.gpu_compat = gpu_compat
-        if gpu_compat is False and theano.config.device == 'gpu':
-            warnings.warn("The function used is not GPU compatible. Please check the gpu_compat flag")
+        if len([i for i in [tau, cov, chol] if i is not None]) != 1:
+            raise ValueError('Incompatible parameterization. Specify exactly '
+                             'one of tau, cov, or chol to specify '
+                             'distribution.')
+        self.mean = self.median = self.mode = self.mu = tt.as_tensor_variable(mu)
+        self.solve = tt.slinalg.Solve(A_structure="lower_triangular", lower=True)
+
+        self.has_tau = tau is not None
+        if cov is not None:
+            self.chol_cov = tt.slinalg.cholesky(tt.as_tensor_variable(cov))
+        elif tau is not None:
+            self.chol_tau = tt.slinalg.cholesky(tt.as_tensor_variable(tau))
+        else:
+            self.chol_cov = tt.as_tensor_variable(chol)
 
     def random(self, point=None, size=None):
-        mu, cov = draw_values([self.mu, self.cov], point=point)
+        if self.has_tau:
+            mu, chol_tau = draw_values([self.mu, self.chol_tau], point=point)
+        else:
+            mu, chol_cov = draw_values([self.mu, self.chol_cov], point=point)
 
-        def _random(mean, cov, size=None):
-            return stats.multivariate_normal.rvs(
-                mean, cov, None if size == mean.shape else size)
+        if size is None:
+            size = []
+        else:
+            try:
+                size = list(size)
+            except TypeError:
+                size = [size]
+        size.append(mu.shape[0])
 
-        samples = generate_samples(_random,
-                                   mean=mu, cov=cov,
-                                   dist_shape=self.shape,
-                                   broadcast_shape=mu.shape,
-                                   size=size)
-        return samples
+        standard_normal = np.random.standard_normal(size)
+        if self.has_tau:
+            return mu + scipy.linalg.solve_triangular(chol_tau, standard_normal.T, lower=True).T
+        return mu + np.dot(standard_normal, chol_cov)
 
     def logp(self, value):
+        if self.has_tau:
+            return self._logp_tau(value)
+        return self._logp_chol(value)
+
+    def _logp_chol(self, value):
         mu = self.mu
-        tau = self.tau
+        delta = value - mu
```

Comment:
Maybe it is easier if we always reshape value to a 2d array. I think `value = value.reshape((-1, n))` should work.